File: agent/storage/store.py
"""JSON-file-backed job and reputation storage.

No DB setup needed for a hackathon build. Not concurrency-hardened beyond
a per-process lock -- fine for the single-process demo runner and for
separate-terminal mode since each store instance re-reads the file on
every op.
"""
from __future__ import annotations
import json
import os
from datetime import datetime, timezone
from threading import Lock

from shared.schema import JobRecord, ReputationRecord
from shared.config import DATA_DIR


class JobStore:
    def __init__(self, path: str | None = None):
        self._path = path or os.path.join(DATA_DIR, "jobs.json")
        self._lock = Lock()

    def _read_all(self) -> dict[str, dict]:
        if not os.path.exists(self._path):
            return {}
        with open(self._path) as f:
            return json.load(f)

    def _write_all(self, data: dict[str, dict]) -> None:
        os.makedirs(os.path.dirname(self._path), exist_ok=True)
        with open(self._path, "w") as f:
            json.dump(data, f, indent=2, default=str)

    def save(self, job: JobRecord) -> None:
        with self._lock:
            data = self._read_all()
            data[job.job_id] = job.model_dump()
            self._write_all(data)

    def get(self, job_id: str) -> JobRecord | None:
        data = self._read_all()
        raw = data.get(job_id)
        return JobRecord(**raw) if raw else None

    def list_all(self) -> list[JobRecord]:
        return [JobRecord(**raw) for raw in self._read_all().values()]


class ReputationStore:
    def __init__(self, path: str | None = None):
        self._path = path or os.path.join(DATA_DIR, "reputation.json")
        self._lock = Lock()

    def _read_all(self) -> dict[str, dict]:
        if not os.path.exists(self._path):
            return {}
        with open(self._path) as f:
            return json.load(f)

    def _write_all(self, data: dict[str, dict]) -> None:
        os.makedirs(os.path.dirname(self._path), exist_ok=True)
        with open(self._path, "w") as f:
            json.dump(data, f, indent=2, default=str)

    def get(self, provider_agent_id: str) -> ReputationRecord:
        data = self._read_all()
        raw = data.get(provider_agent_id)
        if raw:
            return ReputationRecord(**raw)
        return ReputationRecord(provider_agent_id=provider_agent_id)

    def list_all(self) -> list[ReputationRecord]:
        return [ReputationRecord(**raw) for raw in self._read_all().values()]

    def record_job(self, provider_agent_id: str, matches: int, mismatches: int, unverifiable: int) -> ReputationRecord:
        with self._lock:
            data = self._read_all()
            raw = data.get(provider_agent_id)
            rec = ReputationRecord(**raw) if raw else ReputationRecord(provider_agent_id=provider_agent_id)

            rec.total_jobs += 1
            rec.accepted_claims += matches
            rec.mismatched_claims += mismatches
            rec.unverifiable_claims += unverifiable
            checkable = rec.accepted_claims + rec.mismatched_claims
            rec.accuracy_score = round(rec.accepted_claims / checkable, 4) if checkable else 1.0
            rec.last_updated = datetime.now(timezone.utc).isoformat()

            data[provider_agent_id] = rec.model_dump()
            self._write_all(data)
        return rec


job_store = JobStore()
reputation_store = ReputationStore()

File: agent-ts/packages/core/src/store.ts
/**
 * JSON-file-backed job and reputation storage. Ported from
 * agent/storage/store.py.
 *
 * No DB setup needed for a hackathon build. Not concurrency-hardened beyond
 * a simple in-process queue per store -- fine for a single-process demo
 * runner and for separate-terminal mode since each store instance re-reads
 * the file on every op.
 */
import fs from "node:fs";
import path from "node:path";
import { DATA_DIR } from "./config.js";
import { JobRecordSchema, ReputationRecordSchema, type JobRecord, type ReputationRecord } from "./schema.js";

// Simple promise-chain mutex -- Node is single-threaded, so this is enough
// to serialize read-modify-write cycles across concurrent async callers.
function makeMutex() {
  let queue: Promise<unknown> = Promise.resolve();
  return function withLock<T>(fn: () => Promise<T> | T): Promise<T> {
    const run = queue.then(fn);
    queue = run.catch(() => undefined);
    return run;
  };
}

class JobStore {
  private path: string;
  private lock = makeMutex();

  constructor(filePath?: string) {
    this.path = filePath || path.join(DATA_DIR, "jobs.json");
  }

  private readAll(): Record<string, unknown> {
    if (!fs.existsSync(this.path)) return {};
    return JSON.parse(fs.readFileSync(this.path, "utf-8"));
  }

  private writeAll(data: Record<string, unknown>): void {
    fs.mkdirSync(path.dirname(this.path), { recursive: true });
    fs.writeFileSync(this.path, JSON.stringify(data, null, 2));
  }

  save(job: JobRecord): Promise<void> {
    return this.lock(() => {
      const data = this.readAll();
      data[job.job_id] = job;
      this.writeAll(data);
    });
  }

  get(jobId: string): JobRecord | null {
    const data = this.readAll();
    const raw = data[jobId];
    return raw ? JobRecordSchema.parse(raw) : null;
  }

  listAll(): JobRecord[] {
    return Object.values(this.readAll()).map((raw) => JobRecordSchema.parse(raw));
  }
}

class ReputationStore {
  private path: string;
  private lock = makeMutex();

  constructor(filePath?: string) {
    this.path = filePath || path.join(DATA_DIR, "reputation.json");
  }

  private readAll(): Record<string, unknown> {
    if (!fs.existsSync(this.path)) return {};
    return JSON.parse(fs.readFileSync(this.path, "utf-8"));
  }

  private writeAll(data: Record<string, unknown>): void {
    fs.mkdirSync(path.dirname(this.path), { recursive: true });
    fs.writeFileSync(this.path, JSON.stringify(data, null, 2));
  }

  get(providerAgentId: string): ReputationRecord {
    const data = this.readAll();
    const raw = data[providerAgentId];
    return raw ? ReputationRecordSchema.parse(raw) : ReputationRecordSchema.parse({ provider_agent_id: providerAgentId });
  }

  listAll(): ReputationRecord[] {
    return Object.values(this.readAll()).map((raw) => ReputationRecordSchema.parse(raw));
  }

  recordJob(providerAgentId: string, matches: number, mismatches: number, unverifiable: number): Promise<ReputationRecord> {
    return this.lock(() => {
      const data = this.readAll();
      const raw = data[providerAgentId];
      const rec = raw ? ReputationRecordSchema.parse(raw) : ReputationRecordSchema.parse({ provider_agent_id: providerAgentId });

      rec.total_jobs += 1;
      rec.accepted_claims += matches;
      rec.mismatched_claims += mismatches;
      rec.unverifiable_claims += unverifiable;
      const checkableCount = rec.accepted_claims + rec.mismatched_claims;
      rec.accuracy_score = checkableCount ? Math.round((rec.accepted_claims / checkableCount) * 10000) / 10000 : 1.0;
      rec.last_updated = new Date().toISOString();

      data[providerAgentId] = rec;
      this.writeAll(data);
      return rec;
    });
  }
}

export const jobStore = new JobStore();
export const reputationStore = new ReputationStore();
